substrate/validator-sets/pallet/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+      let mut coins = Vec::new();
+      for coin in 0 .. self.coins.0 {
+        coins.push(Coin(coin));
+      }
+
+      let mut participants = Vec::new();
+      for participant in self.participants.clone() {
+        participants.push((participant, self.bond));
+      }
+
+      ValidatorSets::<T>::set(
+        ValidatorSetInstance(Session(0), ValidatorSetIndex(1)),
+        Some(ValidatorSet {
+          bond: self.bond,
+          coins: BoundedVec::try_from(coins).unwrap(),
+          participants: BoundedVec::try_from(participants).unwrap(),
+        }),
+      );
+    }
+  }
+
+  #[pallet::event]
+  #[pallet::generate_deposit(pub(super) fn deposit_event)]
+  pub enum Event<T: Config> {
+    Vote {
+      voter: T::AccountId,
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+      // Amount of votes the key now has
+      votes: u16,
+    },
+    KeyGen {
+      instance: ValidatorSetInstance,
+      coin: Coin,
+      key: Key,
+    },
+  }
+
+  #[pallet::error]
+  pub enum Error<T> {
+    /// Validator Set doesn't exist.
+    NonExistentValidatorSet,
+    /// Non-validator is voting.
+    NotValidator,
+    /// Validator Set already generated keys.
+    AlreadyGeneratedKeys,
+    /// Vvalidator has already voted for these keys.
+    AlreadyVoted,
+  }
+
+  #[pallet::call]
+  impl<T: Config> Pallet<T> {
+    #[pallet::weight(0)] // TODO
+    pub fn vote(
+      origin: OriginFor<T>,
+      index: ValidatorSetIndex,
+      coin: Coin,
+      key: Key,
+    ) -> DispatchResult {
+      let signer = ensure_signed(origin)?;
+      // TODO: Do we need to check key is within bounds?
+
+      // TODO: Get session
+      let session: Session = Session(0);
+
+      // Confirm a key hasn't been set for this set instance
+      let instance = ValidatorSetInstance(session, index);
+      if Keys::<T>::get((instance, coin)).is_some() {
+        Err(Error::<T>::AlreadyGeneratedKeys)?;
+      }
+
+      // Confirm the signer is a validator in the set
+      let set = ValidatorSets::<T>::get(instance).ok_or(Error::<T>::NonExistentValidatorSet)?;
+
+      if set.participants.iter().any(|participant| participant.0 == signer) {
+        Err(Error::<T>::NotValidator)?;
+      }
+
+      // Confirm this signer hasn't already voted for these keys
+      if Voted::<T>::get((&signer, &key)).is_some() {
+        Err(Error::<T>::AlreadyVoted)?;
+      }
+      Voted::<T>::set((&signer, &key), Some(()));
+
+      // Add their vote
+      let votes = VoteCount::<T>::mutate((instance, coin, &key), |value| {
+        *value += 1;
+        *value
+      });
+
+      Self::deposit_event(Event::Vote { voter: signer, instance, coin, key: key.clone(), votes });
+
+      // If we've reached consensus, set the key
+      if usize::try_from(votes).unwrap() == set.participants.len() {
+        Keys::<T>::set((instance, coin), Some(key.clone()));
+        Self::deposit_event(Event::KeyGen { instance, coin, key });
+      }
+
+      Ok(())
```

Comment:
would like to sanity check things in pallets before Ok -- understand this pallet will undergo major refactor so not necessary for this PR.

Reply:
Mind suggesting a few sanity checks?

Reply:
no worries for this PR, let's resolve the functional aspects and have a bigger discussion on this for testnet design.